python/cudf/cudf/core/df_protocol.py

```diff
@@ -0,0 +1,729 @@
+        #     buf = self._col.to_numpy()
+        #     b = bytearray()
+
+        #     # TODO: this for-loop is slow; can be implemented in Cython/C/C++ later
+        #     for i in range(buf.size):
+        #         if type(buf[i]) == str:
+        #             b.extend(buf[i].encode(encoding="utf-8"))
+
+        #     # Convert the byte array to a Pandas "buffer" using a NumPy array as the backing store
+        #     buffer = _CuDFBuffer(np.frombuffer(b, dtype="uint8"))
+
+        #     # Define the dtype for the returned buffer
+        #     dtype = (_k.STRING, 8, "u", "=")  # note: currently only support native endianness
+        else:
+            raise NotImplementedError(f"Data type {self._col.dtype} not handled yet")
+
+        return buffer, dtype
+
+    def _unpackbits(self, myarray, bitorder="big"):
+    
+        bitorder_op = {"big": '(myarray[i / 8] >> (7 - i % 8)) & 1;', 
+                        "little": '(myarray[i / 8] >> (i % 8)) & 1;'}
+        operation = bitorder_op.get(bitorder, None)
+        if operation == None:
+            raise KeyError(f"bitorder must be either 'big' or 'little' not '{bitorder}'")
+        _unpackbits_kernel = _core.ElementwiseKernel(
+        'raw uint8 myarray', 'T unpacked',
+        'unpacked = '+ operation,
+        'unpackbits_kernel'
+        )
+
+        if myarray.dtype != cp.uint8:
+            raise TypeError('Expected an input array of unsigned byte data type')
+
+        unpacked = cp.ndarray((myarray.size * 8), dtype=cp.uint8)
+        return _unpackbits_kernel(myarray, unpacked)
+
+    def _get_validity_buffer(self) -> Tuple[_CuDFBuffer, Any]:
+        """
+        Return the buffer containing the mask values indicating missing data and
+        the buffer's associated dtype.
+
+        Raises RuntimeError if null representation is not a bit or byte mask.
+        """
+        
+        null, invalid = self.describe_null
+        if null == 3:
+            _k = _DtypeKind
+            bitmask = self._unpackbits(cp.asarray(self._col._column.mask), bitorder='little')[:len(self._col)]
+            buffer = _CuDFBuffer(bitmask)
+            dtype = (_k.UINT, 8, "C", "=")
+            return buffer, dtype
+
+        elif null == 1:
+            msg = "This column uses NaN as null so does not have a separate mask"
+        elif null == 0:   
+            msg = "This column is non-nullable so does not have a mask"
+        else:
+            raise NotImplementedError("See self.describe_null")
+
+        raise RuntimeError(msg)
+
+    def _get_offsets_buffer(self) -> Tuple[_CuDFBuffer, Any]:
+        """
+        Return the buffer containing the offset values for variable-size binary
+        data (e.g., variable-length strings) and the buffer's associated dtype.
+
+        Raises RuntimeError if the data buffer does not have an associated
+        offsets buffer.
+        """
+        _k = _DtypeKind
+        if self.dtype[0] == _k.STRING:
```

Comment:
String columns in cuDF store the offsets as one of their children -- can we use that here?

```python

In [9]: df["a"]._column.children
Out[9]:
(<cudf.core.column.numerical.NumericalColumn object at 0x7f7be02c5c20>
 [
   0,
   3,
   7,
   12
 ]
 dtype: int32,
 <cudf.core.column.numerical.NumericalColumn object at 0x7f7be02764d0>
 [
   97,
   98,
   99,
   120,
   121,
   122,
   119,
   108,
   109,
   110,
   111,
   112
 ]
 dtype: int8)

```

Reply:
Yes certainly. But as I've not implemented the `string` support yet. This code should have been commented as it is that of pandas' implementation that I used as an inspiration.